sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala

```diff
@@ -89,9 +89,22 @@ class RelationalGroupedDataset protected[sql](
     case expr: NamedExpression => expr
     case a: AggregateExpression if a.aggregateFunction.isInstanceOf[TypedAggregateExpression] =>
       UnresolvedAlias(a, Some(Column.generateAlias))
-    case expr: Expression => Alias(expr, toPrettySQL(expr))()
+    case ag: UnresolvedFunction if (containsStar(Seq(ag))) || ag.isDistinct =>
+             UnresolvedAlias(expr, None)
+    case expr: Expression =>
+       if (containsStar(Seq(expr))) {
+            UnresolvedAlias(expr, None)
+        } else {
+          Alias(expr, toPrettySQL(expr))()
```

Comment:
If we want a surgical fix, shall we fix how `toPrettySQL` handles star?

Reply:
What should toPrettySQL be generating as alias?  Currently it emits count(unresolvedstar()). I tried count(distinct *) then the query output was:
scala> df.groupBy("id").agg(count_distinct($"*"))
res0: org.apache.spark.sql.DataFrame = [id: bigint, count(distinct *): bigint]
So I am wondering if the fix is in toPrettySQL or the Analyzer. Analyzer should be able to handle unresolvedstar aliases.
The original query does not have an alias for the aggregate expression. For such queries the Origin in Analyzer is (None,,,,..). Whereas if you give a spark.sql query 
spark.sql(" SELECT id, COUNT(DISTINCT *) FROM table GROUP BY id ")
The origin is not none. I wonder if this is what throws off the Analyzer.

Reply:
> If we want a surgical fix, shall we fix how `toPrettySQL` handles star?

How can we fix toPrettySQL to provide the unresolved values of an expr ? The best we can do is to make it count(distinct *) which is still not the same as what spark.sql produces.

Besides it seems like that the alias in case of aggreagate values in dataframe is getting ignored.

scala> df.groupBy("id").agg(count_distinct($"*")).alias("count_alias").show()
+---+-----------------------+                                                   
| id|count(unresolvedstar())|
+---+-----------------------+
|  1|                      1|
|  2|                      1|
|  3|                      1|
|  4|                      1|
|  5|                      1|
|  6|                      1|
|  7|                      1|
|  8|                      1|
|  9|                      1|
+---+-----------------------+

scala> spark.sql(" SELECT id, COUNT(DISTINCT *) as count_alias  FROM table GROUP BY id ").show()
+---+-----------+
| id|count_alias|
+---+-----------+
|  1|          1|
|  2|          1|
|  3|          1|
|  4|          1|
|  5|          1|
|  6|          1|
|  7|          1|
|  8|          1|
|  9|          1|
+---+-----------+

Reply:
> If we want a surgical fix, shall we fix how `toPrettySQL` handles star?

Sure we can fix toPrettySQL. But the best we can do is to get count(distinct *)  which is not the same as what spark.sql  produces.

If we want to duplicate spark.sql behavior the best option would be to create unresolvedAlias for  expr containing "*" as  pushed in the latest build.